### PR TITLE
Warn on use of virtual packages in pkg.installed state

### PR DESCRIPTION
**Pull request by terminalmage**
See https://github.com/saltstack/salt/pull/47263

Note that this will need to be removed once merged forward to develop, I've opened https://github.com/saltstack/salt/issues/47396 to track this task.

### CODE REVIEW

File: salt/states/pkg.py [hunk 1]
```diff
@@ -582,22 +601,28 @@ def _find_install_targets(name=None,
...
                 # The package is installed and is the correct version
-                return {'name': name,
-                        'changes': {},
-                        'result': True,
-                        'comment': 'Version {0} of package \'{1}\' is already '
-                                   'installed'.format(version, name)}
+                ret = {'name': name,
+                       'changes': {},
+                       'result': True,
+                       'comment': 'Version {0} of package \'{1}\' is already '
```
Comment (isbm) [inline salt/states/pkg.py]: BTW, sometimes we use single quotes, sometimes double quotes, sometimes none at all in these situations. Maybe at least we can get rid of those ugly back-slashes in the code. So either one looks OK to me:

```python
'Version {0} of package {1} is already '  # This is your preferred, seems like
```

```python
"Version {0} of package '{1}' is already "
```

```python
'Version {0} of package "{1}" is already '
```

:wink:

File: salt/states/pkg.py [hunk 2]
```diff
@@ -1720,6 +1753,22 @@ def installed(
...
+                    cver = new_pkgs.get(pkgname, [])
+                    if '1' in cver:
+                        virtual_pkgs.append(pkgname)
+                if virtual_pkgs:
+                    warnings.extend(_warn_virtual(virtual_pkgs))
+            except Exception:
+                # This is just some temporary code to warn the user about using
+                # virtual packages. Don't let an exception break the entire
+                # state.
+                pass
```
Comment (isbm) [inline salt/states/pkg.py]: Can we at leaast `log.debug` here this until the code goes away?

**Review by isbm — CHANGES_REQUESTED** [review]

File: salt/states/pkg.py (hunk 2 above)
Comment (terminalmage) [reply salt/states/pkg.py]: Good idea, I've added debug logging here.

**Review by rallytime — APPROVED** [review]

File: salt/states/pkg.py (hunk 1 above)
Comment (isbm) [reply salt/states/pkg.py]: Well, then sounds like  Option 1 is preferred for you indeed.

**Review by isbm — APPROVED** [review]

**Comment by cachedout** [comment]
@terminalmage Quick lint fix requested please: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/21436/violations/file/salt/states/pkg.py/

**Pull request merged**